packages/salesforce-adapter/src/custom_references/rules_and_conditions.ts

```diff
@@ -0,0 +1,208 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  InstanceElement,
+  isInstanceElement,
+  isReferenceExpression,
+  ReferenceInfo,
+} from '@salto-io/adapter-api'
+import { logger } from '@salto-io/logging'
+import { collections, values } from '@salto-io/lowerdash'
+import _ from 'lodash'
+import {
+  CPQ_ADVANCED_CONDITION_FIELD,
+  CPQ_ERROR_CONDITION,
+  CPQ_INDEX_FIELD,
+  CPQ_PRICE_CONDITION,
+  CPQ_PRICE_RULE,
+  CPQ_PRODUCT_RULE,
+  CPQ_QUOTE_TERM,
+  CPQ_QUOTE_TERM_FIELD,
+  CPQ_RULE_FIELD,
+  CPQ_TERM_CONDITON,
+  SBAA_ADVANCED_CONDITION_FIELD,
+  SBAA_APPROVAL_CONDITION,
+  SBAA_APPROVAL_RULE,
+  SBAA_INDEX_FIELD,
+} from '../constants'
+import { WeakReferencesHandler } from '../types'
+
+const log = logger(module)
+const { makeArray } = collections.array
+const { isDefined } = values
+
+type RuleAndConditionDef = {
+  rule: {
+    apiName: string
+    customConditionField: string
+  }
+  condition: {
+    apiName: string
+    indexField: string
+    ruleField: string
+  }
+}
+
+const defs: RuleAndConditionDef[] = [
+  // CPQ Product Rules
+  {
+    rule: {
+      apiName: CPQ_PRODUCT_RULE,
+      customConditionField: CPQ_ADVANCED_CONDITION_FIELD,
+    },
+    condition: {
+      apiName: CPQ_ERROR_CONDITION,
+      indexField: CPQ_INDEX_FIELD,
+      ruleField: CPQ_RULE_FIELD,
+    },
+  },
+  // CPQ Quote Terms
+  {
+    rule: {
+      apiName: CPQ_QUOTE_TERM,
+      customConditionField: CPQ_ADVANCED_CONDITION_FIELD,
+    },
+    condition: {
+      apiName: CPQ_TERM_CONDITON,
+      indexField: CPQ_INDEX_FIELD,
+      ruleField: CPQ_QUOTE_TERM_FIELD,
+    },
+  },
+  // CPQ Price Rules
+  {
+    rule: {
+      apiName: CPQ_PRICE_RULE,
+      customConditionField: CPQ_ADVANCED_CONDITION_FIELD,
+    },
+    condition: {
+      apiName: CPQ_PRICE_CONDITION,
+      indexField: CPQ_INDEX_FIELD,
+      ruleField: CPQ_RULE_FIELD,
+    },
+  },
+  // SBAA Approval Rules
+  {
+    rule: {
+      apiName: SBAA_APPROVAL_RULE,
+      customConditionField: SBAA_ADVANCED_CONDITION_FIELD,
+    },
+    condition: {
+      apiName: SBAA_APPROVAL_CONDITION,
+      indexField: SBAA_INDEX_FIELD,
+      ruleField: SBAA_APPROVAL_RULE,
+    },
+  },
+]
+
+const createReferencesFromRuleInstance = (
+  ruleInstance: InstanceElement,
+  conditionsByIndex: Record<string, InstanceElement>,
+  def: RuleAndConditionDef,
+): ReferenceInfo[] => {
+  const condition = ruleInstance.value[def.rule.customConditionField]
+  if (!_.isString(condition)) {
+    return []
+  }
+  const regexMatch = condition.match(/?\d+/g)
+  if (regexMatch == null) {
+    return []
+  }
+  // We should avoid creating mutliple references to the same condition
+  // e.g. the following custom condition is valid: "0 OR (0 AND 1)"
+  const indexes = _.uniq(Array.from(regexMatch.values()))
+  return indexes
+    .map<ReferenceInfo | undefined>((index) => {
```

Comment:
I think it's more readable to put the type on the function:
```suggestion
    .map((index): ReferenceInfo | undefined => {
```